### ISSUE TITLE
Python bindings use custom type casters instead of pybind11's Eigen support?

### ISSUE THREAD

**Issue by patrikhuber** [libigl/libigl#891]
Hi,

libigl seems to use its own type casters for Eigen matrices and define types like `pyigl.eigen.MatrixXd`. Other types like `Eigen::Vector3d` are not defined.

pybind11 has its own Eigen <=> numpy bindings for quite a while now and it seemlessly converts Eigen matrices to and from numpy arrays, and it works with all Eigen matrix and vector sizes out of the box.

Is there any reason libigl is using these custom bindings and not (the in my opinion much better) pybind11's Eigen <=> numpy support?

**Comment by jdumas** [comment]
We're planning on updating the python bindings at some point to improve that yes. @skoch9 and @fwilliams can tell you more about our plans.

**Comment by fwilliams** [comment]
Hey @patrikhuber. @skoch9 and I have been working on a full rewrite of the Python bindings for libIGL 

As you said, at the moment, the bindings wrap Eigen types directly. They also use output parameters which is extremely un-pythonic. 

The new bindings, will use [NumpyEigen](https://github.com/fwilliams/numpyeigen), a library I wrote for zero-overhead bindings from Numpy/Scipy to Eigen. This means that the new bindings will have full compatibility with Numpy and Scipy with zero copy overhead.  

Furthermore, the function definitions will all become "pythonic", i.e. they will accept inputs, and return a tuple of outputs. The types of outputs will be determined by a `dtype` argument as is the case in scipy. E.g. `V, F = igl.readOBJ("myfile.obj", dtype=(np.float32, np.int32))`. 

We're also building pip and anaconda packages for libIGL which should make it easier to use as a dependency in your project. 

I'm always happy to hear from users, so if you have any feature requests, please let me know and we can discuss them!

**Comment by patrikhuber** [comment]
Hi, cool!

May I ask why you're not just using pybind11's Eigen support? https://pybind11.readthedocs.io/en/master/advanced/cast/eigen.html
They also support passing arguments without copying data and many more things, and the bindings are extremely well maintained.
I don't think they support sparse matrices yet but that would probably not be too hard to add. In contrast to that, from your [example](https://github.com/fwilliams/numpyeigen#example), I think that's extremely cumbersome to use, you basically have to re-write/re-define every function? And you introduce new syntax to learn.
I saw you're still using pybind11 under the hood but I think it would be much better to use pybind11's Eigen bindings and if something is missing then add it there with their help. In that way, lots of other people could benefit too, without having to use your different and separate numpyeigen code. And it would make contributing to libigl easier, because many people can write pybind11 bindings code, but not so many know how to use and write/change code with your library. Just my two cents! :-)

**Comment by fwilliams** [comment]
Thanks for the comment! pybind11's Eigen and Numpy support was the first solution we considered when doing this project. The main reason we chose to implement a binding library was performance. It's a very common pattern to glue together fast pieces of code in a scripting language like Python to build numerical programs. Since glue code is not the bottleneck in these programs, we can build very fast programs in Python. In a fair number of applications we have here in the lab, some libIGL functions get called in a loop with modified input data at every iteration, a pattern which makes data copying undesirable. 

The problem with pybind11 is that there are  some tricky performance caveats. If you allow Numpy arrays of multiple types and orderings (e.g. column major vs. row major), then pybind11 makes a silent copy if the Numpy and the Eigen type don't match (https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html). In real world scenarios, data can come from many different sources and be in different types and layouts. I don't want to introduce performance inconsistencies that are hard to find to a program. libIGL should "just work" well out of the box. 

While its possible to specify multiple types and allow overloading with pybind11, the overloading mechanism is implemented inefficiently: specifically pybind11 implements overloading by calling one version of the function, catching a `type_error` and trying again. If your overload is not the first to be called, you pay a price.

Pybind11 **cannot** get around the above two problems because Eigen requires type information be specified at compile time and Numpy's type information is only available at runtime! Because of this, the only way to implement true fast overloading and zero copy casting is to inspect the input Numpy array metadata, and choose an appropriate type. To make this work with the C++ type system, a user would need to write a big switch statement with a lot of calls to `pybind11:detail::*` to inspect types and do casting correctly. In fact, the first thing I did was write this switch statement by hand! You could add some sugar to make writing this switch statement cleaner but it still winds up being a lot of code and you can still introduce bugs (I did several times when I tried) and copies. 

In essence, NumpyEigen is automatically generating this switch statement for you by using an intermediate transpiler which converts NumpyEigen C++ to C++. The `npe_*`  macros are used to construct an always-correct python binding which ensures that there are no bugs or copy overhead. The code is **drastically** more concise and less error-prone than the hand written alternative with pybind11. 

I've also done a lot of work to make the code generation process transparent: error line numbers are the same after code generation and the code generator catches a lot of errors ahead of time with good messages. The example I have on GitHub is designed to show every feature available in NumpyEigen, which makes it seem more cumbersome than it actually is. 

In summary, NumpyEigen solves a problem that pybind11 cannot solve, and one which I believe is important to make libIGL a fast, user friendly Python library. It does add some overhead but it is much less than writing the correct bindings by hand. The learning curve for NumpyEigen is quite minimal and I recommend giving it a try.
